src/lib.rs

```diff
@@ -128,65 +196,95 @@ impl DstService {
-                            let protocol_hint = if *h2 {
-                                Some(pb::ProtocolHint {
-                                    protocol: Some(pb::protocol_hint::Protocol::H2(
-                                        pb::protocol_hint::H2::default(),
-                                    )),
-                                })
-                            } else {
-                                None
-                            };
-
-                            let mut metric_labels = HashMap::default();
-                            metric_labels.insert("addr".to_string(), addr.to_string());
-                            metric_labels.insert("h2".to_string(), h2.to_string());
-
-                            pb::WeightedAddr {
-                                addr: Some(addr.into()),
-                                protocol_hint,
-                                metric_labels,
-                                ..Default::default()
-                            }
-                        })
-                        .collect::<Vec<_>>();
-                    if !added.is_empty() {
-                        tracing::debug!(?added);
-
-                        let mut metric_labels = HashMap::default();
-                        metric_labels.insert("concrete".to_string(), concrete_name.clone());
-
+                    if curr.is_empty() {
                         tx.send(Ok(pb::Update {
-                            update: Some(pb::update::Update::Add(pb::WeightedAddrSet {
-                                addrs: added,
-                                metric_labels,
+                            update: Some(pb::update::Update::NoEndpoints(pb::NoEndpoints {
+                                exists: true,
                             })),
                         }))
-                        .await?;
-                    }
+                        .await?
+                    } else {
+                        let added = curr
+                            .iter()
+                            .filter(|(_, ep)| Self::is_add(&prev, ep))
+                            .map(|(addr, meta)| {
+                                let protocol_hint = if meta.h2 {
+                                    Some(pb::ProtocolHint {
+                                        protocol: Some(pb::protocol_hint::Protocol::H2(
+                                            pb::protocol_hint::H2::default(),
+                                        )),
+                                    })
+                                } else {
+                                    None
+                                };
+
+                                let mut metric_labels = HashMap::default();
+                                metric_labels.insert("addr".to_string(), addr.to_string());
+                                metric_labels.insert("h2".to_string(), meta.h2.to_string());
+                                metric_labels.extend(meta.metric_labels.clone());
```

Comment:
TIOLI: any reason we can't just start with `meta.metric_labels.clone()`?
```suggestion
                                let mut metric_labels = meta.metric_labels.clone();
                                metric_labels.insert("addr".to_string(), addr.to_string());
                                metric_labels.insert("h2".to_string(), meta.h2.to_string());
```

Reply:
The metric_labels on the EndpointMeta is a BTreeMap because it implementes Hash. but we need to supply a HashMap to the WeightedAddr...